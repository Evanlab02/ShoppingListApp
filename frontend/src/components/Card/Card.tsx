// Third party imports
import { SlIcon } from "@shoelace-style/shoelace/dist/react";

// File imports
import { CardProps } from "./helpers/interfaces";

// Styles
import "./styles/Card.scss"

/**
 * Card component.
 * 
 * @param props 
 * @returns The card component.
 */
export default function Card(props: CardProps) {
    const { backgroundColor, iconBackgroundColor, mainText, subText = "", height = "200px", iconName = "shop", directLink = "", onClick } = props;

    return (
        <div className="card" style={{ backgroundColor: backgroundColor, height: height }}>
            <div className="card-icons">
                <a href={directLink}>
                    <div className="left-icon" style={{ backgroundColor: iconBackgroundColor }}>
                        <SlIcon name={iconName} />
                    </div>
                </a>
                <div className="right-icon"
                    style={{ backgroundColor: iconBackgroundColor }}
                    onClick={() => { if (onClick) { onClick() } }}
                >
                    <SlIcon name="three-dots" />
                </div>
            </div>
            <div className="card-content">
                <h1>{mainText}</h1>
                <h4>{subText}</h4>
            </div>
        </div >
    );
}