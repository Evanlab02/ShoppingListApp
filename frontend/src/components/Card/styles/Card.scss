.card {
    font-family: "Roboto", sans-serif;
    margin: 0.2rem;
    border-radius: 0.5rem;

    .card-icons {
        display: flex;
        align-items: center;
        justify-content: space-between;

        a {
            text-decoration: none;
            cursor: pointer;
            color: #ffffff;

            &:hover,
            &:active,
            &:visited,
            &:focus,
            &:link {
                text-decoration: none;
            }
        }

        .left-icon {
            margin: 1rem 0 0 1rem;
            width: 50px;
            height: 50px;
            border-radius: 0.3rem;
            color: #ffffff;
            display: flex;
            align-items: center;
            justify-content: center;
        }

        .left-icon:hover {
            opacity: 0.5;
        }

        .right-icon {
            margin: 1rem 1rem 0 0;
            width: 50px;
            height: 50px;
            border-radius: 0.3rem;
            color: #ffffff;
            display: flex;
            align-items: center;
            justify-content: center;
        }

        .right-icon:hover {
            cursor: pointer;
            opacity: 0.5;
        }
    }

    .card-content {
        margin: 1rem 0 0 1rem;
        color: #ffffff;
        font-weight: bold;

        h1 {
            margin-bottom: 0.3rem !important;
        }

        h2,
        h4 {
            padding: 0rem;
            margin: 0rem;
        }

        h4 {
            font-weight: 400;
        }
    }
}