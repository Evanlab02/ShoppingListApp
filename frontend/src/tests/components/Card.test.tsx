import { beforeEach, expect, test, vi } from "vitest";
import { fireEvent, render } from "@testing-library/react";
import Card from "../../components/Card/Card";

beforeEach(() => {
	document.getElementsByTagName("html")[0].innerHTML = "";
});

test("Basic Card renders correctly.", async () => {
    const { findByTestId } = render(
        <Card
            backgroundColor="#602786"
            iconBackgroundColor="#562f6f"
            mainText="Test Card"
        />
	);

	const card = await findByTestId("sa-card");
	expect(card).toMatchSnapshot();
});

test("Advanced Card renders correctly.", async () => {
    const fn = vi.fn();

    const { findByTestId } = render(
        <Card
            backgroundColor="#602786"
            iconBackgroundColor="#562f6f"
            mainText="Advanced Card"
            directLink="/hi/"
            iconName="wallet2"
            subText="Hello World!"
            onClick={fn}
        />
	);

	const card = await findByTestId("sa-card");
	expect(card).toMatchSnapshot();

    const click = await findByTestId("sa-card-click");
    fireEvent.click(click);
    expect(fn).toBeCalledTimes(1);
});