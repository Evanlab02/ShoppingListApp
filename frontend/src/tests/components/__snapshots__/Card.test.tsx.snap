// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Advanced Card renders correctly. 1`] = `
<div
  class="card"
  data-testid="sa-card"
  style="background-color: rgb(96, 39, 134); height: 200px;"
>
  <div
    class="card-icons"
  >
    <a
      href="/hi/"
    >
      <div
        class="left-icon"
        style="background-color: rgb(86, 47, 111);"
      >
        <sl-icon
          aria-hidden="true"
          library="default"
        />
      </div>
    </a>
    <div
      class="right-icon"
      data-testid="sa-card-click"
      style="background-color: rgb(86, 47, 111);"
    >
      <sl-icon
        aria-hidden="true"
        library="default"
      />
    </div>
  </div>
  <div
    class="card-content"
  >
    <h1>
      Advanced Card
    </h1>
    <h4>
      Hello World!
    </h4>
  </div>
</div>
`;

exports[`Basic Card renders correctly. 1`] = `
<div
  class="card"
  data-testid="sa-card"
  style="background-color: rgb(96, 39, 134); height: 200px;"
>
  <div
    class="card-icons"
  >
    <a
      href=""
    >
      <div
        class="left-icon"
        style="background-color: rgb(86, 47, 111);"
      >
        <sl-icon
          aria-hidden="true"
          library="default"
        />
      </div>
    </a>
    <div
      class="right-icon"
      data-testid="sa-card-click"
      style="background-color: rgb(86, 47, 111);"
    >
      <sl-icon
        aria-hidden="true"
        library="default"
      />
    </div>
  </div>
  <div
    class="card-content"
  >
    <h1>
      Test Card
    </h1>
    <h4 />
  </div>
</div>
`;
